File: sorting/src/main/java/com/github/solairerove/algs4/leprosorium/sorting/BubbleSort.java
package com.github.solairerove.algs4.leprosorium.sorting;

import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.StdOut;

import java.util.Arrays;

public class BubbleSort {

    public static void sort(Comparable[] a) {
        int n = a.length;

        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (less(a[j + 1], a[j])) {
                    swap(a, j + 1, j);
                }
            }
        }
    }

    private static boolean less(Comparable v, Comparable w) {
        return v.compareTo(w) < 0;
    }

    private static void swap(Comparable[] a, int i, int j) {
        Comparable t = a[i];
        a[i] = a[j];
        a[j] = t;
    }

    private static void show(Comparable[] a) {
        for (Comparable el : a) {
            StdOut.print(el + " ");
        }
        StdOut.println();
    }

    private static boolean isSorted(Comparable[] a) {
        for (int i = 1; i < a.length; ++i) {
            if (less(a[i], a[i - 1])) {
                return false;
            }
        }

        return true;
    }

    public static void main(String... args) {
        String s = new In("/Users/solairerove/Downloads/algs4-data/tiny.txt").readLine();
        String[] array = s.split(" ");

        show(array);
        sort(array);
        show(array);

        int[] a = new int[]{8, 5, 2, 9, 5, 6, 3};
        System.out.println(Arrays.toString(a));
        bubbleSort(a);
        System.out.println(Arrays.toString(a));
    }

    private static void bubbleSort(int[] a) {
        int n = a.length;

        int cnt = 0;
        boolean isSorted = false;
        while (!isSorted) {
            isSorted = true;

            for (int i = 0; i < n - 1 - cnt; i++) {
                if (a[i + 1] < a[i]) {
                    int temp = a[i];
                    a[i] = a[i + 1];
                    a[i + 1] = temp;

                    isSorted = false;
                }
            }
            cnt++;
        }
    }
}

File: src/main/kotlin/com/github/solairerove/algs4/leprosorium/sorting/BubbleSort.kt
package com.github.solairerove.algs4.leprosorium.sorting

fun main() {
    val items = mutableListOf(5, 6, 4, 3, 10, 9, 5, 6, 7)
    print("items: $items \n")
    bubbleSort(items)
    print("items: $items \n")
}

// O(n^2) time | O(1) space
private fun bubbleSort(arr: MutableList<Int>) {
    val n = arr.size

    for (i in 1 until n) {
        for (j in 0 until n - i) {
            if (arr[j + 1] < arr[j]) {
                swap(arr, j + 1, j)
            }
        }
    }
}

private fun swap(arr: MutableList<Int>, i: Int, j: Int) {
    val temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp
}
